File: day10/nim/part1/main.nim
import std/options
from std/strutils import parseInt

let f = open("../../input")

var
  addxVal = none(int)
  cycle = 0
  x = 1
  strengthSum = 0

while true:
  inc cycle
  case cycle
  of 20, 60, 100, 140, 180, 220:
    strengthSum += cycle * x
  else: discard

  if addxVal.isSome:
    x += addxVal.get()
    addxVal = none(int)
    continue

  var line: string
  try:
    line = f.readLine()
  except EOFError:
    break

  var space = line.find(' ')
  if space == -1:
    space = len(line)
  case line[0..space-1]
  of "addx":
    addxVal = some(parseInt(line[space+1..len(line)-1]))
  of "noop": discard
  else:
    raise newException(ValueError, "unexpected instruction")

stdout.write(strengthSum)

File: day10/go/part1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	f, err := os.Open("../../input")
	if err != nil {
		panic(err)
	}

	s := bufio.NewScanner(f)

	var addxV *int
	x, strengthSum := 1, 0

	for cycle := 1; ; cycle++ {
		switch cycle {
		case 20, 60, 100, 140, 180, 220:
			strengthSum += cycle * x
		}

		if addxV != nil {
			x += *addxV
			addxV = nil
			continue
		}

		if !s.Scan() {
			break
		}
		line := s.Text()
		opEnd := strings.IndexByte(line, ' ')
		if opEnd == -1 {
			opEnd = len(line)
		}

		switch line[:opEnd] {
		case "addx":
			i, err := strconv.Atoi(line[opEnd+1:])
			if err != nil {
				panic(err)
			}
			addxV = &i
		case "noop":
		default:
			panic("unexpected instruction")
		}
	}
	if s.Err() != nil {
		panic(err)
	}

	fmt.Print(strengthSum)
}
